keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,674 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+    - [ ] e2e Tests for all Beta API Operations (endpoints)
+    - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+    - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+    - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal aims at enabling hot plugging of node compute resources. This will help in updating cluster resource capacity by just resizing compute resources of nodes rather than adding new node to a cluster.
+The updated node configurations are to be reflected at the node and cluster levels automatically without the need to reset the kubelet.
+
+This proposal also aims to improve the initialization and reinitialization of resource managers, such as the CPU manager and memory manager, in response to changes in a node's CPU and memory configurations.
+
+## Motivation
+In a typical Kubernetes environment, the cluster resources may need to be altered due to following reasons:
+- Incorrect resource assignment during cluster creation.
+- Increased workload over time, leading to the need for additional resources in the cluster.
+
+To handle these scenarios, we can:
+- Horizontally scale up the cluster by adding compute nodes.
+- Vertically scale up the cluster by increasing node capacity. However, currently, the workaround for capturing node resizing in the cluster involves restarting the Kubelet.
+
+Node resource hot plugging will provide advantages in scenarios such as:
+- Handling resource demand with a limited set of nodes by increasing the capacity of existing nodes instead of creating new nodes.
+- Creating new nodes takes more time compared to increasing the capacity of existing nodes.
+
+### Goals
```

Comment:
what about NRI plugins? Handling any other sideeffects?

Reply:
I'd say out of scope, though it is relevant to NRI plugin authors that the values could change and they should be reactive to that.

Reply:
I do not think that assessing impact of this KEP on components, especially known components like NRI plugins can be out of scope. Saying that we ignore any issues caused by that and we will document it is a form of assessment and it should be included.

Specifically for NRI plugins, one alternative we can (should?) discuss is whether we need to notify the container runtime about the change in resources how kubelet sees them.

Reply:
Since the runtimes (and by extension NRI) don't get initialyy notified by the kubelet on what the kubelet thinks of the nodes resources, I'd expect whatever mechanism they use to gather the node stats can be updated to handle hot plug in the same way kubelet is. Unless we want to introduce a resource API that can be called up into, I don't think we should notify down in this way. Though, calling it out as such I think is a good idea

Reply:
@Karthik-K-N have you reviewed that neither arguments of existing CRI APIs depend on the current value of the available memory and cpu? If so than we just need to raise awareness via documentation. I guess cpu quotas are only depends on the processor so one more processor wouldn't matter when we pass CPU quota to the runtime.

I can only think of the memory swap that is dependent on the total memory on the machine. Is there a separtate thread discussing how we deal with it?

Reply:
yes, only Memory swap and OOMScoreAdj depends on total memory on the machine. We are planning to update them for existing container via `UpdateContainerResources` CRI API.

Also can you please  help by pointing to document which needs to be updated to add these information?

Reply:
> Since the runtimes (and by extension NRI) don't get initialyy notified by the kubelet on what the kubelet thinks of the nodes resources,

Have to agree here. We never defined the node resource API or even try to ensure the available resources seen by the runtime is consistent with what kubelet sees. It's strange for kubelet to suddenly send a resource update to the runtime without knowing what the call means.

Reply:
https://github.com/kubernetes/enhancements/pull/5319 it seems this KEP is planning to introduce a node resource API coming from CRI, and implies it's a prereq for this KEP

Reply:
No, these two KEPs are independent. #3953 (i.e. this one) makes the node resources hot-pluggable but the data still comes from cAdvisor. #5319 proposes to change the source of truth from cAdvisor to something else. #5319 on its own would get the info from non-cAdvisor but only at kubelet startup (no hotplug). Makes sense?